tests/Unit/Evolution/DiscontinuousGalerkin/Test_ConservationHelper.cpp

```diff
@@ -0,0 +1,167 @@
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Helpers/DataStructures/MakeWithRandomValues.hpp"
+#include "Helpers/Evolution/DiscontinuousGalerkin/BoundaryCorrections.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/Formulation.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace {
+namespace Tags {
+struct Var1 : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+
+template <size_t Dim>
+struct Var2 : db::SimpleTag {
+  using type = tnsr::i<DataVector, Dim, Frame::Inertial>;
+};
+
+struct VolumeDouble : db::SimpleTag {
+  using type = double;
+};
+}  // namespace Tags
+
+template <size_t Dim>
+struct System {
+  static constexpr bool is_in_flux_conservative_form = true;
+  static constexpr bool has_primitive_and_conservative_vars = false;
+  static constexpr size_t volume_dim = Dim;
+
+  using variables_tag =
+      ::Tags::Variables<tmpl::list<Tags::Var1, Tags::Var2<Dim>>>;
+  using flux_variables = tmpl::list<Tags::Var1, Tags::Var2<Dim>>;
+  using gradient_variables = tmpl::list<>;
+  using sourced_variables = tmpl::list<>;
+
+  struct TimeDerivativeTerms {
+    using temporary_tags = tmpl::list<>;
+  };
+
+  using compute_volume_time_derivative_terms = TimeDerivativeTerms;
+};
+
+template <size_t Dim>
+struct Correction {
+ private:
+  struct AbsCharSpeed : db::SimpleTag {
+    using type = Scalar<DataVector>;
+  };
+
+ public:
+  using dg_package_field_tags =
+      tmpl::list<Tags::Var1, ::Tags::NormalDotFlux<Tags::Var1>, Tags::Var2<Dim>,
+                 ::Tags::NormalDotFlux<Tags::Var2<Dim>>, AbsCharSpeed>;
+  using dg_package_data_temporary_tags = tmpl::list<>;
+  using dg_package_data_volume_tags = tmpl::list<Tags::VolumeDouble>;
+
+  double dg_package_data(
+      const gsl::not_null<Scalar<DataVector>*> packaged_var1,
+      const gsl::not_null<Scalar<DataVector>*> packaged_normal_dot_flux_var1,
+      const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+          packaged_var2,
+      const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+          packaged_normal_dot_flux_var2,
+      const gsl::not_null<Scalar<DataVector>*> packaged_abs_char_speed,
+      const Scalar<DataVector>& var1,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>& var2,
+      const tnsr::I<DataVector, Dim, Frame::Inertial>& flux_var1,
+      const tnsr::Ij<DataVector, Dim, Frame::Inertial>& flux_var2,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_covector,
+      const boost::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+      /*mesh_velocity*/,
+      const boost::optional<Scalar<DataVector>>& normal_dot_mesh_velocity,
+      const double volume_double) const noexcept {
+    *packaged_var1 = var1;
+    *packaged_var2 = var2;
+    dot_product(packaged_normal_dot_flux_var1, flux_var1, normal_covector);
+    for (size_t i = 0; i < Dim; ++i) {
```

Comment:
is it deliberate to calculate the dot products in two different ways?

Reply:
Yea... `dot_production` is (vector, vector) -> (scalar) so doesn't work in the second case